File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> chars;
        for (char i: s) {
            if (i == ')') {
                if (chars.size() > 0 && chars.top() == '(') {
                    chars.pop();
                } else return false;
            } else if (i == ']') {
                if (chars.size() > 0 && chars.top() == '[') {
                    chars.pop();
                } else return false;
            } else if (i == '}') {
                if (chars.size() > 0 && chars.top() == '{') {
                    chars.pop();
                } else return false;
            } else {
                if (i == '[' || i == '(' || i == '{') {
                    chars.push(i);
                } else return false;
            }
        }

        return chars.size() == 0;
    }
};

File: LeetCode/Easy/0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        valids = {'}': '{', ')': '(', ']': '['}
        cur = []

        for ch in s:
            if ch not in valids:
                cur.append(ch)
            else:
                if len(cur) == 0:
                    return False
                
                if cur[-1] == valids[ch]:
                    cur.pop()
                else:
                    return False


        return len(cur) == 0